pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -49,103 +55,78 @@ type forwarder struct {
+	// request and response represent the processors used to handle
+	// client-to-server and server-to-client messages.
+	request  processor
+	response processor
+
+	// errCh is a buffered channel that contains the first forwarder error.
+	// This channel may receive nil errors. When an error is written to this
+	// channel, it is guaranteed that the forwarder and all connections will
+	// be closed.
+	errCh chan error
 }
 
 // forward returns a new instance of forwarder, and starts forwarding messages
 // from clientConn to serverConn. When this is called, it is expected that the
-// caller passes ownership of serverConn to the forwarder, which implies that
-// the forwarder will clean up serverConn. clientConn and serverConn must not
-// be nil in all cases except for testing.
+// caller passes ownership of both clientConn and serverConn to the forwarder,
+// which implies that the forwarder will clean them up. clientConn and
+// serverConn must not be nil in all cases except for testing.
 //
-// Note that callers MUST call Close in all cases, even if ctx was cancelled.
+// Note that callers MUST call Close in all cases, even if ctx was cancelled,
+// and callers will need to detect that (for now).
 func forward(ctx context.Context, clientConn, serverConn net.Conn) *forwarder {
 	ctx, cancelFn := context.WithCancel(ctx)
-
 	f := &forwarder{
-		ctx:       ctx,
-		ctxCancel: cancelFn,
-		errChan:   make(chan error, 1),
+		ctx:        ctx,
+		ctxCancel:  cancelFn,
+		errCh:      make(chan error, 1),
+		clientConn: interceptor.NewBackendConn(clientConn),
+		serverConn: interceptor.NewFrontendConn(serverConn),
 	}
-
-	// The net.Conn object for the client is switched to a net.Conn that
-	// unblocks Read every second on idle to check for exit conditions.
-	// This is mainly used to unblock the request processor whenever the
-	// forwarder has stopped, or a transfer has been requested.
-	clientConn = pgwire.NewReadTimeoutConn(clientConn, func() error {
-		// Context was cancelled.
-		if f.ctx.Err() != nil {
-			return f.ctx.Err()
-		}
-		// TODO(jaylim-crl): Check for transfer state here.
-		return nil
-	})
-	f.clientConn = interceptor.NewBackendConn(clientConn)
-	f.serverConn = interceptor.NewFrontendConn(serverConn)
-
-	// Start request (client to server) and response (server to client)
-	// processors. We will copy all pgwire messages/ from client to server
-	// (and vice-versa) until we encounter an error or a shutdown signal
-	// (i.e. context cancellation).
-	go func() {
-		defer f.Close()
-
-		err := wrapClientToServerError(f.handleClientToServer())
-		select {
-		case f.errChan <- err: /* error reported */
-		default: /* the channel already contains an error */
-		}
-	}()
-	go func() {
-		defer f.Close()
-
-		err := wrapServerToClientError(f.handleServerToClient())
-		select {
-		case f.errChan <- err: /* error reported */
-		default: /* the channel already contains an error */
-		}
-	}()
-
+	f.request = &requestProcessor{f: f}
```

Comment:
There _will be_ shared state in the future (with the connection migration PR). Eventually this forwarder field will come back if we remove it now. Perhaps you can also argue that the shared state doesn't need to be stored in the forwarder (i.e. it can be in another struct). Regardless, I can add a TODO comment to revisit this.